node/src/main/scala/coop/rchain/node/web/Transaction.scala

```diff
@@ -0,0 +1,234 @@
+
+final case class TransactionInfo(transaction: Transaction, transactionType: TransactionType)
+
+final case class TransactionResponse(data: Seq[TransactionInfo])
+
+trait TransactionAPI[F[_]] {
+  def getTransaction(blockHash: String): F[Seq[TransactionInfo]]
+}
+
+/**
+  * This API is totally based on how RevVault.rho is written. If the `RevVault.rho` is re-written or changed,
+  * this API might end up with useless.
+  */
+final case class TransactionAPIImpl[F[_]: Sync: Concurrent](
+    blockReportAPI: BlockReportAPI[F],
+    // The transferUnforgeable can be retrieved based on the deployer and the timestamp of RevVault.rho
+    // in the genesis ceremony.
+    transferUnforgeable: Par
+) extends TransactionAPI[F] {
+  private def findUserDeployTransaction(
+      d: DeployInfoWithEventData,
+      report: SingleReport,
+      index: Int
+  ) = {
+    val transactions = findTransactions(report)
+    index match {
+      // the index order is based on how we play a user deploy
+      case 0 =>
+        transactions.map(t => TransactionInfo(t, PreCharge(d.deployInfo.get.sig))).pure
+      case 1 =>
+        transactions.map(t => TransactionInfo(t, UserDeploy(d.deployInfo.get.sig))).pure
+      case 2 =>
+        transactions.map(t => TransactionInfo(t, Refund(d.deployInfo.get.sig))).pure
+      case _ =>
+        Sync[F].raiseError[Vector[TransactionInfo]](
+          new Exception(
+            "It is not possible that user deploy got more than 3 reports(precharge, user, refund)."
+          )
+        )
+    }
+  }
+
+  def getTransaction(blockHash: String): F[Seq[TransactionInfo]] =
+    for {
+      report <- blockReportAPI.blockReport(blockHash, forceReplay = false)
+      ts <- report match {
+             case Right(b) =>
+               for {
+                 udts <- b.deploys.toList.traverse(d => {
+                          d.report.zipWithIndex.toList.traverse {
+                            case (report, index) => findUserDeployTransaction(d, report, index)
+                          }
+                        })
+                 userTransactions = udts.flatten.flatten
+                 systemDeployTransactions = b.systemDeploys.flatMap { s =>
+                   // system doesn't get precharge and refund , so it would always get one
+                   val transactions = findTransactions(s.report.head)
+                   s.systemDeploy.get.systemDeploy.value match {
+                     case SlashSystemDeployDataProto(_, _) =>
+                       transactions.map(
+                         t => TransactionInfo(t, SlashingDeploy(blockHash))
+                       )
+                     case CloseBlockSystemDeployDataProto() =>
+                       transactions.map(t => TransactionInfo(t, CloseBlock(blockHash)))
+                   }
+                 }
+                 transactions = userTransactions ++ systemDeployTransactions
+               } yield transactions
```

Comment:
Let's make this more functional. Here are some hints for user deploy but similarly can be done for system also.
```scala
// Helper which accepts TransactionType constructor
private def makeDeployTransaction(
    d: DeployInfoWithEventData,
    report: SingleReport,
    txCtor: String => TransactionType
) = findTransactions(report).map(t => TransactionInfo(t, txCtor(d.deployInfo.get.sig)))

// List of TransactionType constructors that we want to zip with reports
val txCtors = Seq(PreCharge, UserDeploy, Refund)

// For Either, instead of matching directly on constructors we can use `map` and `getOrElse`
report map { b =>
  // ...
}
.getOrElse(Seq.empty)
```